src/H5FD.c

```diff
@@ -701,66 +701,77 @@ H5FDopen(const char *name, unsigned flags, hid_t fapl_id, haddr_t maxaddr)
 
-    FUNC_ENTER_NOAPI(NULL)
+    FUNC_ENTER_NOAPI(FAIL)
 
     /* Sanity checks */
     if (0 == maxaddr)
-        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, NULL, "zero format address range");
+        HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "zero format address range");
 
     /* Get file access property list */
     if (NULL == (plist = (H5P_genplist_t *)H5I_object(fapl_id)))
-        HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, NULL, "not a file access property list");
+        HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL, "not a file access property list");
 
     /* Get the VFD to open the file with */
     if (H5P_peek(plist, H5F_ACS_FILE_DRV_NAME, &driver_prop) < 0)
-        HGOTO_ERROR(H5E_PLIST, H5E_CANTGET, NULL, "can't get driver ID & info");
+        HGOTO_ERROR(H5E_PLIST, H5E_CANTGET, FAIL, "can't get driver ID & info");
 
     /* Get driver info */
     if (NULL == (driver = (H5FD_class_t *)H5I_object(driver_prop.driver_id)))
-        HGOTO_ERROR(H5E_VFL, H5E_BADVALUE, NULL, "invalid driver ID in file access property list");
+        HGOTO_ERROR(H5E_VFL, H5E_BADVALUE, FAIL, "invalid driver ID in file access property list");
     if (NULL == driver->open)
-        HGOTO_ERROR(H5E_VFL, H5E_UNSUPPORTED, NULL, "file driver has no `open' method");
+        HGOTO_ERROR(H5E_VFL, H5E_UNSUPPORTED, FAIL, "file driver has no `open' method");
 
     /* Query driver flag */
     if (H5FD_driver_query(driver, &driver_flags) < 0)
-        HGOTO_ERROR(H5E_VFL, H5E_BADVALUE, NULL, "can't query VFD flags");
+        HGOTO_ERROR(H5E_VFL, H5E_BADVALUE, FAIL, "can't query VFD flags");
 
     /* Get initial file image info */
     if (H5P_peek(plist, H5F_ACS_FILE_IMAGE_INFO_NAME, &file_image_info) < 0)
-        HGOTO_ERROR(H5E_PLIST, H5E_CANTGET, NULL, "can't get file image info");
+        HGOTO_ERROR(H5E_PLIST, H5E_CANTGET, FAIL, "can't get file image info");
 
     /* If an image is provided, make sure the driver supports this feature */
-    assert(((file_image_info.buffer != NULL) && (file_image_info.size > 0)) ||
-           ((file_image_info.buffer == NULL) && (file_image_info.size == 0)));
-    if ((file_image_info.buffer != NULL) && !(driver_flags & H5FD_FEAT_ALLOW_FILE_IMAGE))
-        HGOTO_ERROR(H5E_VFL, H5E_UNSUPPORTED, NULL, "file image set, but not supported.");
+    assert((file_image_info.buffer && file_image_info.size > 0) ||
+           (!file_image_info.buffer && file_image_info.size == 0));
+    if (file_image_info.buffer && !(driver_flags & H5FD_FEAT_ALLOW_FILE_IMAGE))
+        HGOTO_ERROR(H5E_VFL, H5E_UNSUPPORTED, FAIL, "file image set, but not supported.");
 
-    /* Dispatch to file driver */
     if (HADDR_UNDEF == maxaddr)
         maxaddr = driver->maxaddr;
-    if (NULL == (file = (driver->open)(name, flags, fapl_id, maxaddr)))
-        HGOTO_ERROR(H5E_VFL, H5E_CANTINIT, NULL, "open failed");
+
+    /* Try dispatching to file driver */
+    H5E_PAUSE_ERRORS
+    {
+        file = (driver->open)(name, flags, fapl_id, maxaddr);
```

Comment:
Shouldn't this function have some method of telling it whether or not to pause errors when dispatching to the file driver? Otherwise, since several places like `H5F_open` are being switch to using this new function, it's just always going to hide errors when file opens fail.

Reply:
Yes, that's true - it always hides errors now, because of the changes to the H5F code.  The previous H5F code was already accommodating those failures by clearing the stack when it failed.   This change, and the changes to the H5F code, moves that down here and accommodates for the change here with more straightforward logic in the H5F code.

Reply:
I get that one would want to hide errors for speculative file open attempts, but it seems like code (like `H5F_open`) that was previously performing both a speculative and non-speculative file open has been changed to only make use of these try_open calls, in essence making two speculative file open attempts.

Reply:
I should probably clarify that I'm talking about errors from the VOL/VFD level being hidden, not necessarily errors from the library's default error stack. For example, running a simple program that tries to open an HDF5 file which doesn't exist before and after these changes gives:

```C
#include "hdf5.h"

int
main(int argc, char **argv)
{
    H5Fopen("nonexist.h5", H5F_ACC_RDWR, H5P_DEFAULT);
    return 0;
}
```

Before
```
HDF5-DIAG: Error detected in HDF5 (1.15.0):
  #000: /home/jhenderson/git/hdf5/src/H5F.c line 827 in H5Fopen(): unable to synchronously open file
    major: File accessibility
    minor: Unable to open file
  #001: /home/jhenderson/git/hdf5/src/H5F.c line 788 in H5F__open_api_common(): unable to open file
    major: File accessibility
    minor: Unable to open file
  #002: /home/jhenderson/git/hdf5/src/H5VLcallback.c line 3689 in H5VL_file_open(): open failed
    major: Virtual Object Layer
    minor: Can't open object
  #003: /home/jhenderson/git/hdf5/src/H5VLcallback.c line 3501 in H5VL__file_open(): open failed
    major: Virtual Object Layer
    minor: Can't open object
  #004: /home/jhenderson/git/hdf5/src/H5VLnative_file.c line 128 in H5VL__native_file_open(): unable to open file
    major: File accessibility
    minor: Unable to open file
  #005: /home/jhenderson/git/hdf5/src/H5Fint.c line 1901 in H5F_open(): unable to open file: name = 'nonexist.h5', tent_flags = 1
    major: File accessibility
    minor: Unable to open file
  #006: /home/jhenderson/git/hdf5/src/H5FD.c line 763 in H5FD_open(): open failed
    major: Virtual File Layer
    minor: Unable to initialize object
  #007: /home/jhenderson/git/hdf5/src/H5FDsec2.c line 323 in H5FD__sec2_open(): unable to open file: name = 'nonexist.h5', errno = 2, error message = 'No such file or directory', flags = 1, o_flags = 2
    major: File accessibility
    minor: Unable to open file
```

After
```
HDF5-DIAG: Error detected in HDF5 (1.15.0):
  #000: /home/jhenderson/HDF5_Workspace/Worktrees/test_error/src/H5F.c line 827 in H5Fopen(): unable to synchronously open file
    major: File accessibility
    minor: Unable to open file
  #001: /home/jhenderson/HDF5_Workspace/Worktrees/test_error/src/H5F.c line 788 in H5F__open_api_common(): unable to open file
    major: File accessibility
    minor: Unable to open file
  #002: /home/jhenderson/HDF5_Workspace/Worktrees/test_error/src/H5VLcallback.c line 3689 in H5VL_file_open(): open failed
    major: Virtual Object Layer
    minor: Can't open object
  #003: /home/jhenderson/HDF5_Workspace/Worktrees/test_error/src/H5VLcallback.c line 3501 in H5VL__file_open(): open failed
    major: Virtual Object Layer
    minor: Can't open object
  #004: /home/jhenderson/HDF5_Workspace/Worktrees/test_error/src/H5VLnative_file.c line 128 in H5VL__native_file_open(): unable to open file
    major: File accessibility
    minor: Unable to open file
  #005: /home/jhenderson/HDF5_Workspace/Worktrees/test_error/src/H5Fint.c line 2247 in H5F_open(): can't open file
    major: File accessibility
    minor: Unable to open file
```